connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -582,4 +625,159 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
             messageBatch.clear();
         }
     }
+
+    static class SinkTaskMetricsGroup {
+        private final MetricGroup metricGroup;
+        private final Sensor sinkRecordRead;
+        private final Sensor sinkRecordSend;
+        private final Sensor sinkRecordLag;
+        private final Sensor partitionCount;
+        private final Sensor offsetSeqNum;
+        private final Sensor offsetCompletion;
+        private final Sensor offsetCompletionSkip;
+        private final Sensor putBatchTime;
+        private Map<TopicPartition, OffsetAndMetadata> consumedOffsets = new HashMap<>();
+        private Map<TopicPartition, OffsetAndMetadata> processedOffsets = new HashMap<>();
+
+        public SinkTaskMetricsGroup(ConnectorTaskId id, ConnectMetrics connectMetrics) {
+            metricGroup = connectMetrics.group("sink-task-metrics",
+                    "connector", id.connector(), "task", Integer.toString(id.task()));
+
+            sinkRecordRead = metricGroup.metrics().sensor("sink-record-read");
+            sinkRecordRead.add(metricGroup.metricName("sink-record-read-rate",
+                    "The average per-second number of records read from Kafka for this task belonging to the " +
+                            "named sink connector in this worker. This is before transformations are applied."),
+                    new Rate());
+            sinkRecordRead.add(metricGroup.metricName("sink-record-read-total",
+                    "The total number of records produced/polled (before transformation) by this task belonging " +
+                            "to the named sink connector in this worker, since the task was last restarted."),
+                    new Total());
+
+            sinkRecordSend = metricGroup.metrics().sensor("sink-record-send");
+            sinkRecordSend.add(metricGroup.metricName("sink-record-send-rate",
+                    "The average per-second number of records output from the transformations and sent/put to " +
+                            "this task belonging to the named sink connector in this worker. This is after transformations " +
+                            "are applied and excludes any records filtered out by the transformations."),
+                    new Rate());
+            sinkRecordSend.add(metricGroup.metricName("sink-record-send-total",
+                    "The total number of records output from the transformations and sent/put to this task " +
+                            "belonging to the named sink connector in this worker, since the task was last restarted."),
+                    new Total());
+
+            sinkRecordLag = metricGroup.metrics().sensor("sink-record-lag");
+            sinkRecordLag.add(metricGroup.metricName("sink-record-lag-max",
+                    "The maximum lag in terms of number of records that the sink task is behind the consumer's position for any " +
```

Comment:
Ok, sorry, I'm thinking more about this now with review, and I guess this will always just be either 0 or 1 batch of messages since the processing -> put() will be synchronous for each batch collected from the consumer.

So I guess maybe the committed - consumed makes sense as it is the total still thought to be somewhere in flight (or more accurately, not yet known to be guaranteed delivered into the destination) does actually work. I think, as you mentioned, lag is just confusing there because you could be completely done processing, the data could be in the destination, and we may just not yet have gotten to a periodic commit yet. I mainly would worry about that since connect defaults don't commit all that frequently and it is hard to say what it means if, e.g., the HDFS connector returns a large "lag" since it *needs* large "lag" to write large files.

:( sorry, i think this might need some more thought

Reply:
The number of records being _processed_ should always be less than or equal to the batch size; the worst case scenario is when the consumer has read the full batch from a single topic partition, and in this case the processing lag would be equal to the batch size.

However, with connectors like the HDFS connector where commit is a more critical stage of processing, the commit lag seems to be far more useful and important: how many records have been consumed but not yet _written completely_ (committed) to the sink system. 

I've got the changes locally to revert back to the _committed lag_, so just let me know what you prefer.

Reply:
right, i may also be playing fast and loose w/ terminology in my responses :) But I think we're on the same page as to the meaning of the different options.

My concern with the HDFS uncommitted case is that compared to consumer lag where, at least roughly, you could pick a number and use it across all topics for alerting, this doesn't have that property. The value you would alert on for HDFS is completely different from the value you'd alert on for ES. In this way "lag" definitely seems inaccurate terminology -- the connector isn't really "behind", it just intentionally has uncommitted data.

I see a few options:

* Include both! Something like "processed-lag" and "committed-lag".
* Don't use "lag" when referring to uncommitted data. It could just be something like "uncommitted-count" (which then opens the question of doing it per partition, max across partitions, or sum over all partitions).

And of course these aren't necessarily mutually exclusive. The potential drawbacks I see with the first point is that maybe the processed lag just isn't that interesting since it is bounded by the # of messages that fit in a single fetch request anyway, and introducing both may lead to confusion if we're not super clear in the docs about which is more important to monitor. In particular, now that I've thought through it more I'm struggling to find a problem I could discover / diagnose based on the processed lag.

The drawback to the second is that we're diverging from the terminology used elsewhere. But maybe that's fine since, as we're discovering through this conversation, they are in fact different things.

You could always re-raise this in the KIP discussion thread to get additional feedback. Also, might be worth looking at whether streams has any relevant metrics and how they handled this. In some ways they have the same problem -- they have the actual consumer lag, but also the time for a message to make it through the topology and all downstream messages to be acked. At best I would guess they could track lineage within a subtopology and provide info based on that, but I am guessing those metrics may simply not be there due to the more complex processing graph that they have.

Reply:
oh, one other thing I just thought of -- if you are filtering messages aggressively with an smt, will metrics like committed-lag become misleading since they may not advance very frequently?

Reply:
Streams doesn't seem to track any kind of lag, and I would expect that can only be computed by comparing consumer and producer offsets. And that works because the streams are always writing back to topics.

Connectors, of course, have to track how many records are in-flight in the connector, before being written to Kafka (for source connectors) or after having read from Kafka (for sink connectors). It seems like what we want to measure, then, is the number of in-flight records, or IOW, the max & avg number of records that the connector is "holding" at any point in time.

So, how about the following new metrics for source connectors:
* `source-record-process-queue-count-max` - the maximum number of source records that the source task had created but had not yet been written to Kafka or discarded by the transformations
* `source-record-process-queue-count-avg` - the average number of source records that the source task had created but had not yet been written to Kafka or discarded by the transformations

and the following new metrics for sink connectors:
* `sink-record-process-queue-count-max` - the maximum number of sink records that have been read from Kafka and not yet transformed and processed by the the sink task
* `sink-record-process-queue-count-avg` - the average number of sink records that have been read from Kafka and not yet transformed and processed by the the sink task
* `sink-record-flush-queue-count-max` - the maximum number of sink records that have been processed but not yet flushed by the sink task
* `sink-record-flush-queue-count-avg` - the average number of sink records that have been processed but not yet flushed by the sink task

And regarding "lag", isn't it true that since the consumer offsets are committed _after_ the sink task has flushed the records that the normal consumer lag actually tells you how far behind the sink task is? Sure, it's a bit larger than actual lag for connectors like HDFS that don't really do anything on flush, but that seems a relatively smaller problem than the fact that the offsets are committed infrequently.

WDYT?